packages/artifact/src/internal/requestUtils.ts

```diff
@@ -0,0 +1,74 @@
+import {IHttpClientResponse} from '@actions/http-client/interfaces'
+import {isRetryableStatusCode, isSuccessStatusCode, sleep} from './utils'
+import * as core from '@actions/core'
+
+export async function retry<T>(
+  name: string,
+  method: () => Promise<T>,
+  getStatusCode: (response: T) => number | undefined,
+  errorMessages: Map<number, string>,
+  maxAttempts: number,
+  delay: number
+): Promise<T> {
+  let response: T | undefined = undefined
+  let statusCode: number | undefined = undefined
+  let isRetryable = false
+  let errorMessage = ''
+  let extraErrorInformation: string | undefined = undefined
+  let attempt = 1
+
+  while (attempt <= maxAttempts) {
+    try {
+      response = await method()
+      statusCode = getStatusCode(response)
+
+      if (isSuccessStatusCode(statusCode)) {
+        return response
+      }
+
+      // Extra error information that we want to display if a particular response code is hit
+      if (statusCode) {
+        extraErrorInformation = errorMessages.get(statusCode)
+      }
+
+      isRetryable = isRetryableStatusCode(statusCode)
```

Comment:
I think you can take this a step further to separate this into an HTTP layer and an HTTP-agnostic `retry` function:
- Pull out `getStatusCode`, `isSuccessStatusCode`, `isRetryableStatusCode` from the `retry` function and replace it with a single function you pass in called `shouldRetry`
- `retryHttpClientResponse` then can take the HTTP-related stuff and reduce it to a `shouldRetry` callback

Reply:
I was mostly going off of the existing `retry` function that `cache` uses. I feel like pulling out `isSuccessStatusCode` and `getStatusCode` could cause some repetition throughout the rest of the code. Overall I think the current pattern is sufficient for the calls that we need to make.

Reply:
It shouldn't cause any repetition because `retryHttpClientRequest` will shim it.  It might work but it's not very easy to understand right now.  Separation of concerns will fix with that.

Reply:
Spent a bit of time trying to pull out each of the methods, but I could really get it in a nice enough format without breaking too much and I don't think it works all that well. The current `getStatusCode` method is actually structured the way it is so that it's easier to test each of the responses: 

https://github.com/actions/toolkit/blob/1bcfb42ea312c2640cde7d04a81b6c03c19025b5/packages/artifact/__tests__/retry.test.ts#L55

My original intention was to use mostly what `cache` had without changing it too much so that if there is a fix in one package the two are relatively the same and it's easy to follow.